File: introduction/ITP1_9_D-Transformation.js
var s = require('fs').readFileSync('/dev/stdin', 'utf8').trim().split('\n');
for(var i = 0; i < s[1]; i++) {
  var o = s[i+2].split(' ');
  var a = parseInt(o[1]), b = parseInt(o[2])+1;
  switch(o[0]) {
    case 'print':
      console.log(s[0].slice(a, b));
      break;
    case 'reverse':
      s[0] = s[0].slice(0, a)+s[0].slice(a, b).split('').reverse().join('')+s[0].slice(b);
      break;
    case 'replace':
      s[0] = s[0].slice(0, a)+o[3]+s[0].slice(b);
      break;
  }
}

File: introduction/ITP1_9_D-Transformation.scala
object Main extends App {
  val line = io.Source.stdin.getLines.toList
  var s = line(0)
  line.drop(2).map(_.split(" ").toList).map(o => {
    val a = o(1).toInt
    val b = o(2).toInt+1
    val as = s slice(0, a)
    val bs = s slice(b, s.length)
    o(0) match {
      case "print" => println(s slice(a,b))
      case "reverse" => s = as+(s slice(a,b)).reverse+bs
      case "replace" => s = as+o(3)+bs
    }
  })
}
